File: src/main/java/org/arbor/gtnn/data/GTNNMaterials.java
package org.arbor.gtnn.data;

import com.gregtechceu.gtceu.GTCEu;
import com.gregtechceu.gtceu.api.data.chemical.material.Material;
import com.gregtechceu.gtceu.api.data.chemical.material.info.MaterialIconSet;
import com.gregtechceu.gtceu.api.data.chemical.material.properties.FluidProperty;
import com.gregtechceu.gtceu.api.data.chemical.material.properties.OreProperty;
import com.gregtechceu.gtceu.api.data.chemical.material.properties.PropertyKey;
import com.gregtechceu.gtceu.api.fluids.FluidBuilder;
import com.gregtechceu.gtceu.api.fluids.store.FluidStorageKeys;
import org.arbor.gtnn.GTNNIntegration;
import org.arbor.gtnn.config.ConfigHandler;
import org.arbor.gtnn.data.materials.*;

import static com.gregtechceu.gtceu.api.data.chemical.material.info.MaterialIconSet.SHINY;

public class GTNNMaterials {
    // MaterialIconSet
    public static MaterialIconSet InfinityIcon = new MaterialIconSet("infinity", SHINY);
    // region first degree mats
    public static Material AndesiteAlloy;
    public static Material Desh;
    public static Material Ostrum;
    public static Material Calorite;
    public static Material SpaceNeutronium;
    public static Material Infinity;
    public static Material InfinityCatalyst;
    // Chemical Plant
    public static Material RP1;
    public static Material RP1RocketFuel;
    public static Material Kerosene;
    public static Material DenseHydrazineMixedFuel;
    public static Material Hydrazine;
    public static Material HydrogenPeroxide;
    public static Material EthylAnthraQuinone;
    public static Material EthylAnthraHydroQuinone;
    public static Material Anthracene;
    public static Material MethylhydrazineNitrateRocketFuel;
    public static Material MethylHydrazine;
    public static Material UDMHRocketFuel;
    public static Material UDMH;
    public static Material OrangeMetalCatalyst;
    public static Material PhthalicAnhydride;
    public static Material VanadiumPentoxide;
    public static Material BlackMatter;
    public static Material Cerrobase140;
    public static Material SodiumPyrosulfate;
    public static Material SodiumSulfate;
    public static Material ZincSulfate;
    public static Material Wollastonite;
    public static Material ArcaneCrystal;
    // Ingot
    public static Material ManaSteel;
    public static Material TerraSteel;
    public static Material Elementium;
    public static Material RefinedRadiance;
    public static Material ShadowSteel;
    // Platinum
    public static Material PlatinumSalt;
    public static Material RefinedPlatinumSalt;
    public static Material PalladiumSalt;
    public static Material RhodiumNitrate;
    public static Material RoughlyRhodiumMetal;
    public static Material PalladiumMetal;
    public static Material MetalSludge;
    public static Material PlatinumSlag;
    public static Material ReprecipitatedRhodium;
    public static Material SodiumNitrate;
    public static Material RhodiumSalt;
    public static Material RhodiumFilterCake;
    public static Material PlatinumMetal;
    public static Material Kaolinite;
    public static Material Dolomite;
    public static Material SodiumRutheniate;
    public static Material IridiumDioxide;
    public static Material ConcentratedPlatinum;
    public static Material PalladiumRichAmmonia;
    public static Material RutheniumTetroxideLQ;
    public static Material SodiumFormate;
    public static Material FormicAcid;
    public static Material RhodiumSulfateGas;
    public static Material AcidicIridium;
    public static Material RutheniumTetroxideHot;
    // Naquadah
    public static Material NaquadahOxideMixture;
    public static Material EnrichedNaquadahOxideMixture;
    public static Material NaquadriaOxideMixture;
    public static Material HexafluorideEnrichedNaquadahSolution;
    public static Material XenonHexafluoroEnrichedNaquadate;
    public static Material PalladiumOnCarbon;
    public static Material GoldTrifluoride;
    public static Material EnrichedNaquadahResidueSolution;
    public static Material XenoauricFluoroantimonicAcid;
    public static Material GoldChloride;
    public static Material BromineTrifluoride;
    public static Material HexafluorideNaquadriaSolution;
    public static Material RadonDifluoride;
    public static Material RadonNaquadriaOctafluoride;
    public static Material NaquadriaResidueSolution;
    public static Material CaesiumFluoride;
    public static Material XenonTrioxide;
    public static Material CaesiumXenontrioxideFluoride;
    public static Material NaquadriaCaesiumXenonnonfluoride;
    public static Material RadonTrioxide;
    public static Material NaquadriaCaesiumfluoride;
    public static Material NitrosoniumOctafluoroxenate;
    public static Material NitrylFluoride;
    public static Material AcidicNaquadriaCaesiumfluoride;

    public static void init() {
        FirstMaterials.init();
        SecondMaterials.init();
        if (ConfigHandler.INSTANCE.Server.enableHarderPlatinumLine) PlatinumLineMaterials.init();
        if (ConfigHandler.INSTANCE.Server.enableHarderNaquadahLine) NaquadahMaterials.init();
        if (GTNNIntegration.isAdAstraLoaded()) AdAstraMaterials.init();
        if (GTNNIntegration.isBotaniaLoaded()) BotaniaMaterials.init();
        if (GTNNIntegration.isCreateLoaded()) CreateMaterials.init();
        if (GTNNIntegration.isForbiddenArcanusLoaded()) ForbiddenArcanusMaterials.init();
    }

    public static void addFluid(Material material){
        material.setProperty(PropertyKey.FLUID, new FluidProperty());
        material.getProperty(PropertyKey.FLUID).getStorage().enqueueRegistration(FluidStorageKeys.LIQUID, new FluidBuilder());
    }

    public static void addOre(Material material){
        material.setProperty(PropertyKey.ORE, new OreProperty());
    }

    public static Material.Builder Builder(String id) {
        return new Material.Builder(GTCEu.id(id));
    }
}

File: src/main/java/dev/arbor/gtnn/data/GTNNMaterials.kt
package dev.arbor.gtnn.data

import com.gregtechceu.gtceu.GTCEu
import com.gregtechceu.gtceu.api.data.chemical.material.Material
import com.gregtechceu.gtceu.api.data.chemical.material.info.MaterialIconSet
import com.gregtechceu.gtceu.api.data.chemical.material.properties.DustProperty
import com.gregtechceu.gtceu.api.data.chemical.material.properties.FluidProperty
import com.gregtechceu.gtceu.api.data.chemical.material.properties.OreProperty
import com.gregtechceu.gtceu.api.data.chemical.material.properties.PropertyKey
import com.gregtechceu.gtceu.api.fluids.FluidBuilder
import com.gregtechceu.gtceu.api.fluids.store.FluidStorageKeys
import dev.arbor.gtnn.GTNN.getServerConfig
import dev.arbor.gtnn.GTNNIntegration.isAdAstraLoaded
import dev.arbor.gtnn.GTNNIntegration.isBotaniaLoaded
import dev.arbor.gtnn.GTNNIntegration.isCreateLoaded
import dev.arbor.gtnn.api.recipe.GTNNBuilder
import dev.arbor.gtnn.data.materials.*
import dev.arbor.gtnn.data.recipes.BrineChain

object GTNNMaterials {
    lateinit var AndesiteAlloy: Material
    lateinit var Desh: Material
    lateinit var Ostrum: Material
    lateinit var Calorite: Material
    lateinit var SpaceNeutronium: Material
    lateinit var Infinity: Material
    lateinit var InfinityCatalyst: Material
    lateinit var RP1: Material
    lateinit var RP1RocketFuel: Material
    lateinit var Kerosene: Material
    lateinit var DenseHydrazineMixedFuel: Material
    lateinit var Hydrazine: Material
    lateinit var HydrogenPeroxide: Material
    lateinit var EthylAnthraQuinone: Material
    lateinit var EthylAnthraHydroQuinone: Material
    lateinit var Anthracene: Material
    lateinit var MethylhydrazineNitrateRocketFuel: Material
    lateinit var MethylHydrazine: Material
    lateinit var UDMHRocketFuel: Material
    lateinit var UDMH: Material
    lateinit var OrangeMetalCatalyst: Material
    lateinit var PhthalicAnhydride: Material
    lateinit var VanadiumPentoxide: Material
    lateinit var BlackMatter: Material
    lateinit var Cerrobase140: Material
    lateinit var PotassiumPyrosulfate: Material
    lateinit var SodiumSulfate: Material
    lateinit var ZincSulfate: Material
    lateinit var Wollastonite: Material
    lateinit var ArcaneCrystal: Material
    lateinit var ManaSteel: Material
    lateinit var TerraSteel: Material
    lateinit var Elementium: Material
    lateinit var RefinedRadiance: Material
    lateinit var ShadowSteel: Material
    lateinit var PlatinumSalt: Material
    lateinit var RefinedPlatinumSalt: Material
    lateinit var PalladiumSalt: Material
    lateinit var RhodiumNitrate: Material
    lateinit var RoughlyRhodiumMetal: Material
    lateinit var PalladiumMetal: Material
    lateinit var MetalSludge: Material
    lateinit var PlatinumSlag: Material
    lateinit var ReprecipitatedRhodium: Material
    lateinit var SodiumNitrate: Material
    lateinit var RhodiumSalt: Material
    lateinit var RhodiumFilterCake: Material
    lateinit var PlatinumMetal: Material
    lateinit var Kaolinite: Material
    lateinit var Dolomite: Material
    lateinit var SodiumRutheniate: Material
    lateinit var IridiumDioxide: Material
    lateinit var ConcentratedPlatinum: Material
    lateinit var PalladiumRichAmmonia: Material
    lateinit var RutheniumTetroxideLQ: Material
    lateinit var SodiumFormate: Material
    lateinit var FormicAcid: Material
    lateinit var RhodiumSulfateGas: Material
    lateinit var AcidicIridium: Material
    lateinit var RutheniumTetroxideHot: Material
    lateinit var NaquadahOxideMixture: Material
    lateinit var EnrichedNaquadahOxideMixture: Material
    lateinit var NaquadriaOxideMixture: Material
    lateinit var HexafluorideEnrichedNaquadahSolution: Material
    lateinit var XenonHexafluoroEnrichedNaquadate: Material
    lateinit var PalladiumOnCarbon: Material
    lateinit var GoldTrifluoride: Material
    lateinit var EnrichedNaquadahResidueSolution: Material
    lateinit var XenoauricFluoroantimonicAcid: Material
    lateinit var GoldChloride: Material
    lateinit var BromineTrifluoride: Material
    lateinit var HexafluorideNaquadriaSolution: Material
    lateinit var RadonDifluoride: Material
    lateinit var RadonNaquadriaOctafluoride: Material
    lateinit var NaquadriaResidueSolution: Material
    lateinit var CaesiumFluoride: Material
    lateinit var XenonTrioxide: Material
    lateinit var CaesiumXenontrioxideFluoride: Material
    lateinit var NaquadriaCaesiumXenonnonfluoride: Material
    lateinit var RadonTrioxide: Material
    lateinit var NaquadriaCaesiumfluoride: Material
    lateinit var NitrosoniumOctafluoroxenate: Material
    lateinit var NitrylFluoride: Material
    lateinit var AcidicNaquadriaCaesiumfluoride: Material
    lateinit var GraphiteUraniumMixture: Material
    lateinit var UraniumCarbideThoriumMixture: Material
    lateinit var PlutoniumOxideUraniumMixture: Material
    lateinit var Thorium232: Material
    lateinit var ThoriumBasedLiquidFuelExcited: Material
    lateinit var ThoriumBasedLiquidFuelDepleted: Material
    lateinit var ThoriumBasedLiquidFuel: Material
    lateinit var UraniumBasedLiquidFuelExcited: Material
    lateinit var UraniumBasedLiquidFuelDepleted: Material
    lateinit var UraniumBasedLiquidFuel: Material
    lateinit var PlutoniumBasedLiquidFuelExcited: Material
    lateinit var PlutoniumBasedLiquidFuelDepleted: Material
    lateinit var PlutoniumBasedLiquidFuel: Material
    lateinit var RadiationProtection: Material
    lateinit var NaquadahBasedLiquidFuel: Material
    lateinit var NaquadahBasedLiquidFuelExcited: Material
    lateinit var NaquadahBasedLiquidFuelDepleted: Material
    lateinit var IodizedBrine: Material
    lateinit var IodineBrineMixture: Material
    lateinit var BrominatedBrine: Material
    lateinit var IodineSlurry: Material
    lateinit var AcidicBrominatedBrine: Material
    lateinit var BromineSulfateSolution: Material
    lateinit var OverheatedBromineSulfateSolution: Material
    lateinit var WetBromine: Material
    lateinit var DebrominatedWater: Material
    lateinit var NeutroniumMixture: Material

    fun addDust(material: Material) {
        material.setProperty(PropertyKey.DUST, DustProperty())
    }

    fun addFluid(material: Material) {
        material.setProperty(PropertyKey.FLUID, FluidProperty())
        material.getProperty(PropertyKey.FLUID).storage.enqueueRegistration(FluidStorageKeys.LIQUID, FluidBuilder())
    }

    fun addGas(material: Material) {
        material.setProperty(PropertyKey.FLUID, FluidProperty())
        material.getProperty(PropertyKey.FLUID).storage.enqueueRegistration(FluidStorageKeys.GAS, FluidBuilder())
    }

    fun addOre(material: Material) {
        material.setProperty(PropertyKey.ORE, OreProperty())
    }

    fun builder(id: String?): GTNNBuilder {
        return GTNNBuilder(GTCEu.id(id))
    }

    fun init() {
        AdjustGTMaterials.init()
        FirstMaterials.init()
        SecondMaterials.init()
        BrineChain.init()
        if (getServerConfig().enableHarderPlatinumLine) PlatinumLineMaterials.init()
        if (getServerConfig().enableHarderNaquadahLine) NaquadahMaterials.init()
        if (isAdAstraLoaded()) AdAstraMaterials.init()
        if (isBotaniaLoaded()) BotaniaMaterials.init()
        if (isCreateLoaded()) CreateMaterials.init()
    }


    object MaterialIcons {
        var InfinityIcon = MaterialIconSet("infinity", MaterialIconSet.SHINY)
    }
}
